indico/modules/events/registration/google_wallet.py

```diff
@@ -0,0 +1,356 @@
+        return wrapper
+    return decorator
+
+
+class GoogleWalletManager:
+    """Google Wallet ticketing manager.
+
+    This class handles the integration with the Google Wallet ticketing API.
+    """
+
+    def __init__(self, event):
+        self.event = event
+        self.settings = self.get_google_wallet_settings(self.event.category)
+        self.credentials, self.http_client = self._load_credentials()
+
+    def _load_credentials(self):
+        """Load Google credentials and setup authenticated HTTP client."""
+        if not self.configured:
+            return None, None
+        credentials = Credentials.from_service_account_info(
+            self.settings['google_wallet_credentials'],
+            scopes=['https://www.googleapis.com/auth/wallet_object.issuer']
+        )
+        http_client = AuthorizedSession(credentials)
+        return credentials, http_client
+
+    @property
+    def configured(self):
+        return config.ENABLE_GOOGLE_WALLET and bool(self.settings['google_wallet_mode'])
+
+    @property
+    def class_id(self):
+        """The unique identifier for an event ticket class.
+
+        Each event has one ticket class that holds shared data such as the title
+        and dates of the event.
+        """
+        return event_wallet_settings.get(self.event, 'google_wallet_class_id', f'Indico-Event-{self.event.id}')
+
+    @classmethod
+    def verify_credentials(cls, account_info_json, issuer_id) -> GoogleCredentialValidationResult:
+        try:
+            credentials = Credentials.from_service_account_info(
+                account_info_json,
+                scopes=['https://www.googleapis.com/auth/wallet_object.issuer']
+            )
+        except GoogleAuthError as exc:
+            logger.info('Could not load GCP credentials: %s', exc)
+            return GoogleCredentialValidationResult.invalid
+        client = AuthorizedSession(credentials)
+        resp = client.get(f'{API_BASE_URL}/issuer/{issuer_id}')
+        if resp.ok:
+            return GoogleCredentialValidationResult.ok
+        logger.info('Could not validate GCP credentials (%d): %s', resp.status_code, resp.text)
+        if resp.status_code == 403:
+            return GoogleCredentialValidationResult.refused
+        elif resp.status_code == 404:
+            return GoogleCredentialValidationResult.bad_issuer
+        else:
+            return GoogleCredentialValidationResult.failed
+
+    @staticmethod
+    def event_uses_google_wallet_tickets(event):
+        return event_wallet_settings.get(event, 'google_wallet_class_id') is not None
+
+    @staticmethod
+    def get_google_wallet_settings(category):
+        while category:
+            if category.google_wallet_settings.get('google_wallet_mode') is not None:
+                return category.google_wallet_settings
+            else:
+                category = category.parent
+        return {'google_wallet_mode': False}
```

Comment:
Certainly not a property since this can trigger multiple DB queries while climbing up the tree (but not enough to justify using CTE magic to do it in a single query). `category.lookup_google_wallet_settings()` would be more appropriate, even though I kind of like having most of the GW stuff in here and not cluttered throughout the various parts of Indico except where it cannot be avoided...

Reply:
moved to a `category.effective_google_wallet_config` CTE in the end (ie a single query regardless of how deep the category tree nesting is)